src/sinks/nats.rs

```diff
@@ -0,0 +1,152 @@
+        &self,
+        _cx: SinkContext,
+    ) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let sink = NatsSink::new(self.clone());
+        let healthcheck = future::ok(()).boxed();
+
+        Ok((super::VectorSink::Stream(Box::new(sink)), healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "nats"
+    }
+}
+
+/**
+ * Code dealing with the Sink struct.
+ */
+
+#[derive(Clone)]
+pub struct NatsSink {
+    nc: nats::Connection,
+    subject: String,
+}
+
+impl NatsSink {
+    fn new(config: NatsSinkConfig) -> Self {
+        let options = nats::Options::new();
+        let nc = options.connect(&config.url).unwrap();
+
+        Self {
+            nc,
+            subject: config.subject,
+        }
+    }
+}
+
+#[async_trait]
+impl StreamSink for NatsSink {
+    async fn run(&mut self, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
+        while let Some(event) = input.next().await {
+            match event {
+                Event::Log(log) => {
+                    let body = log
+                        .get(crate::config::log_schema().message_key())
+                        .map(|v| v.to_string_lossy())
+                        .unwrap_or_else(|| "".into());
+
+                    let message_len = body.len();
+
+                    if self.nc.publish(&self.subject, body).is_ok() {
+                        emit!(NatsEventSent {
+                            byte_size: message_len,
+                        });
+                    }
+                }
+                Event::Metric(_metric) => {}
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(feature = "nats-integration-tests")]
+#[cfg(test)]
+mod integration_tests {
+    use super::*;
+    use crate::test_util::{random_lines_with_stream, random_string, trace_init};
+    use std::{
+        sync::{Arc, Mutex},
+        time::Duration,
+    };
+
+    #[tokio::test]
+    async fn nats_happy() {
+        // Publish `N` messages to NATS.
+        //
+        // Observe with a second subscriber that at least some of the messages
+        // were received.
+        //
+        // NATS operates with at_most_once delivery semantics
+        // - https://docs.nats.io/faq#does-nats-guarantee-message-delivery
+        //
+        // All messages should be accountable in the local integration test
+        // case, but, to prevent flakiness initially, validate the basic
+        // acceptable outcome.
+
+        trace_init();
+
+        let subject = format!("test-{}", random_string(10));
+
+        let cnf = NatsSinkConfig {
+            url: "nats://127.0.0.1:4222".to_owned(),
+            subject: subject.clone(),
+        };
+
+        // Establish the consumer subscription.
+        let consumer = NatsSink::new(cnf.clone());
+        let num_recv_events: Arc<Mutex<u32>> = Arc::new(Mutex::new(0));
+        let sub = consumer.nc.subscribe(&subject).unwrap();
```

Comment:
Nit: Could we just create a standalone `nats` subscriber? It feels a little strange to create the sink just for the side-effect of the consumer. You could move the client creation into a function if you wanted to share it.